File: benchmarks/oop_benchmark/oop.js
class Circle {
  constructor(radius) {
    this.radius = radius;
  }
  area() {
    return Math.trunc((314 * this.radius * this.radius) / 100);
  }
  perimeter() {
    return Math.trunc((628 * this.radius) / 100);
  }
}

class Rectangle {
  constructor(width, height) {
    this.width = width;
    this.height = height;
  }
  area() {
    return this.width * this.height;
  }
  perimeter() {
    return 2 * (this.width + this.height);
  }
}

const N = 100000;
let totalArea = 0;
let totalPerimeter = 0;

for (let i = 0; i < N; i++) {
  const r = 1 + (i % 100);

  if (i % 3 === 0) {
    const c = new Circle(r);
    totalArea += c.area();
    totalPerimeter += c.perimeter();
  } else if (i % 3 === 1) {
    const rect = new Rectangle(r, r * 2);
    totalArea += rect.area();
    totalPerimeter += rect.perimeter();
  } else {
    const sq = new Rectangle(r, r);
    totalArea += sq.area();
    totalPerimeter += sq.perimeter();
  }
}

console.log(totalArea);
console.log(totalPerimeter);

File: benchmarks/oop_benchmark/oop.go
package main

import "fmt"

type Circle struct {
	Radius int
}

func (c *Circle) Area() int {
	return 314 * c.Radius * c.Radius / 100
}

func (c *Circle) Perimeter() int {
	return 628 * c.Radius / 100
}

type Rectangle struct {
	Width  int
	Height int
}

func (r *Rectangle) Area() int {
	return r.Width * r.Height
}

func (r *Rectangle) Perimeter() int {
	return 2 * (r.Width + r.Height)
}

func main() {
	N := 100000
	totalArea := 0
	totalPerimeter := 0

	for i := 0; i < N; i++ {
		r := 1 + i%100

		if i%3 == 0 {
			c := &Circle{Radius: r}
			totalArea += c.Area()
			totalPerimeter += c.Perimeter()
		} else if i%3 == 1 {
			rect := &Rectangle{Width: r, Height: r * 2}
			totalArea += rect.Area()
			totalPerimeter += rect.Perimeter()
		} else {
			sq := &Rectangle{Width: r, Height: r}
			totalArea += sq.Area()
			totalPerimeter += sq.Perimeter()
		}
	}

	fmt.Println(totalArea)
	fmt.Println(totalPerimeter)
}
